taichi/backends/opengl/opengl_api.cpp

```diff
@@ -345,41 +341,33 @@ OpenGlRuntime::~OpenGlRuntime() = default;
 
 void DeviceCompiledProgram::launch(RuntimeContext &ctx,
                                    OpenGlRuntime *runtime) const {
-  std::array<void *, taichi_max_num_args> ext_arr_host_ptrs;
-
   uint8_t *args_buf_mapped = nullptr;
 
+  // Prepare external arrays/ndarrays
+  // Each external array is copied into its own device buffer (SSBO).
+  // - ctx.args[i] contains its ptr on host.
+  // - arr_bufs_[i] contains its DeviceAllocation on device.
+  // Note shapes of these external arrays still reside in argument buffer,
+  // see more details below.
+  for (const auto &[i, size] : program_.ext_arr_map) {
+    if (size == 0)
+      continue;
+    void *baseptr = device_->map(arr_bufs_[i]);
```

Comment:
Later on, I think a hack fix is to recreate a new SSBO here when the size has changed. WDYT?

Reply:
Yup that'll work! But IIUC as you suggested, letting ndarray manage device allocation will also fix the correctness issue (in a less hacky way)? So I'm currently hacking around to get that one works :D

Reply:
Oh yeah, I just felt that the hacky one is less effort to get things working... Feel free to implement the correct one :-)